server/github/client.go

```diff
@@ -0,0 +1,89 @@
+package github
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/Thesohan/weaveGitHubSearchService/server/constants"
+	httpClient "github.com/Thesohan/weaveGitHubSearchService/server/http_client"
+	"github.com/Thesohan/weaveGitHubSearchService/server/structures"
+)
+
+// GitHubAPIURL is the base URL for GitHub search API.
+const GitHubAPIURL = "https://api.github.com/search/code"
+
+var (
+	once      sync.Once
+	singleton IGithubClient
+)
+
+type githubClient struct {
+	client  *http.Client
+	baseURL string
+	headers map[string]string
+}
+
+// IGithubClient interface for API requests.
+type IGithubClient interface {
+	SearchCode(ctx context.Context, query string) (*structures.Data, error)
+}
+
+/*
+	NewGitHubClient returns a singleton instance of GitHubClient.
+
+baseURL is accepted here to make testing easy, we can pass a mock URL while creating tests
+*/
+func NewGitHubClient(baseURL ...string) IGithubClient {
+	once.Do(func() {
+		fmt.Println("Creating new github client")
+		url := GitHubAPIURL
+		if len(baseURL) > 0 && baseURL[0] != "" {
+			url = baseURL[0]
+		}
+		singleton = &githubClient{
+			client:  &http.Client{},
```

Comment:
Never make a go http client without setting a timeout
```suggestion
			client:  &http.Client{Timeout: time.Second *3},
```